Sources/BigInt/BigInt.swift

```diff
@@ -330,37 +330,29 @@ extension BigInt: SignedNumeric {
 extension BigInt: BinaryInteger {
 
   public init?<T>(exactly source: T) where T: BinaryFloatingPoint {
-    if (source.isNaN || source.isInfinite) ||
-      (source.rounded(.towardZero) != source) {
+    guard source.isFinite, source == source.rounded(.towardZero) else {
       return nil
     }
-
     self.init(source)
   }
 
   public init<T>(_ source: T) where T: BinaryFloatingPoint {
-    precondition(
-      !(source.isNaN || source.isInfinite),
-      "\(type(of: source)) value cannot be converted to BigInt because it is either infinite or NaN"
-    )
-
-    let isNegative = source < 0.0
-    var float = isNegative ? -source : source
+    precondition(source.isFinite, "BigInt(\(source)) failed")
+    precondition(T.significandBitCount < .max)
```

Comment:
`BinaryFloatingPoint` defines `significandBitCount` as `Int` specifically, so this condition seems unnecessary? Certainly nothing further down is going to break if someone made a `BinaryFloatingPoint` type with `Int.max` bits in the significand, given that all the shifts are done on `BigInt`s.

Reply:
`Int.max` means "[no upper limit][]" for extensible floating-point types.

This precondition is probably unnecessary for BinaryFloatingPoint types?

[no upper limit]: <https://github.com/apple/swift/blob/5a75996d1d915cfa6c3ce51643bddf02ffd16f97/stdlib/public/core/FloatingPoint.swift#L1571-L1574>

Reply:
I think it's unnecessary, and specifically it would seem we ought to be able to handle the "no upper limit" case since `BigInt` also has no upper limit! I think handling it would require some version of the older division-based code? We could detect `Int.max` here and use that code, otherwise use the bit-twiddling version?